src/gtk/window.cpp

```diff
@@ -2828,6 +2893,661 @@ static gboolean source_dispatch(GSource*, GSourceFunc, void*)
+
+            if ( data->m_touchCount == 1 )
+            {
+                data->m_lastTouchTime = gdk_event->time;
+                data->m_lastTouchPoint.x = gdk_event->x;
+                data->m_lastTouchPoint.y = gdk_event->y;
+
+                // Save the sequence which identifies touch corresponding to "press"
+                data->m_touchSequence = gdk_event->sequence;
+
+                // "Press and Tap Event" may occur in future
+                data->m_allowedGestures |= press_and_tap;
+            }
+
+            // Check if two fingers are placed together .i.e difference between their time stamps is <= 200 milliseconds
+            else if ( data->m_touchCount == 2 && gdk_event->time - data->m_lastTouchTime <= wxTwoFingerTimeInterval )
+            {
+                // "Two Finger Tap Event" may be possible in the future
+                data->m_allowedGestures |= two_finger_tap;
+
+                // Cancel "Press and Tap Event"
+                data->m_allowedGestures &= ~press_and_tap;
+            }
+            break;
+
+        case GDK_TOUCH_UPDATE:
+            // If press and tap gesture is active and touch corresponding to that gesture is moving
+            if ( (data->m_activeGestures & press_and_tap) && gdk_event->sequence == data->m_touchSequence )
+            {
+                data->m_gestureState = update;
+                wxEmitPressAndTapEvent(gdk_event, win);
+            }
+            break;
+
+        case GDK_TOUCH_END:
+        case GDK_TOUCH_CANCEL:
+            data->m_touchCount--;
+
+            if ( data->m_touchCount == 1 )
+            {
+                data->m_lastTouchTime = gdk_event->time;
+
+                // If the touch corresponding to "press" is present and "tap" is produced by some ather touch
+                if ( (data->m_allowedGestures & press_and_tap) && gdk_event->sequence != data->m_touchSequence )
+                {
+                    // Press and Tap gesture becomes active now
+                    if ( !(data->m_activeGestures & press_and_tap) )
+                    {
+                        data->m_gestureState = begin;
+                        data->m_activeGestures |= press_and_tap;
+                    }
+
+                    else
+                    {
+                        data->m_gestureState = update;
+                    }
+
+                    wxEmitPressAndTapEvent(gdk_event, win);
+                }
+            }
+
+            // Check if "Two Finger Tap Event" is possible and both the fingers have been lifted up together
+            else if ( (data->m_allowedGestures & two_finger_tap) && !data->m_touchCount
+                      && gdk_event->time - data->m_lastTouchTime <= wxTwoFingerTimeInterval )
+            {
+                // Process Two Finger Tap Event
+                wxEmitTwoFingerTapEvent(gdk_event, win);
+            }
+
+            // If the gesture was active and the touch corresponding to "press" is no longer on the screen
```

Comment:
wxEmitTwoFingerTapEvent will call a used event handler, which can disable touch events. In this cases, wxWindowGesturesData is freed in the disable touch events functionl. Returning here, data is still at the old value pointing at the freed memory. The next statements will use the memory at data.

Reply:
Thanks, I missed this one. It could be fixed by redoing `wxWindowGestures::FromObject()` but this is definitely brittle. An obvious solution would be to use reference counting, but I wonder if it's really worth the extra complexity. And in the worst case, it's relatively trivial to use `CallAfter()` to postpone the call to `EnableTouchEvents()` slightly.

Is it really realistic to disable touch events from a touch event handler?

Reply:
Or just keep the datastructure for the whole object lifetime and delete it only in the destructor.

Reply:
Yes, this is probably the best/simplest solution, thanks! I'll update the PR soon.